### PR TITLE
drivers: sensor: add driver supporting am2301b

### PR DESCRIPTION
**Pull request by yamaoku-009**
Add driver for aosong am2301b temprature and humidity module.

Product Homepage:
http://www.aosong.com/en/products-64.html

DataSheet:
https://www.aosong.com/userfiles/files/media/AM2301B%E8%8B%B1%E6%96%87%E7%89%88_%E5%8E%BB%E6%BB%A4%E6%B3%A2%20%E4%B8%8A%E6%8B%89.pdf

The pull request previously submitted in #44389 has been closed, so I'm resubmitting it.

### CODE REVIEW

File: drivers/sensor/CMakeLists.txt [hunk 1]
```diff
@@ -111,6 +111,7 @@ add_subdirectory_ifdef(CONFIG_ADC_CMP_NPCX	nuvoton_adc_cmp_npcx)
 add_subdirectory_ifdef(CONFIG_TACH_IT8XXX2	ite_tach_it8xxx2)
 add_subdirectory_ifdef(CONFIG_VCMP_IT8XXX2	ite_vcmp_it8xxx2)
 add_subdirectory_ifdef(CONFIG_PCNT_ESP32	pcnt_esp32)
+add_subdirectory_ifdef(CONFIG_AM2301B		am2301b)
```
Comment (MaureenHelm) [inline drivers/sensor/CMakeLists.txt:114]: Please move this up to for alpha order

File: drivers/sensor/Kconfig [hunk 2]
```diff
@@ -258,4 +258,6 @@ source "drivers/sensor/ite_vcmp_it8xxx2/Kconfig"
 
 source "drivers/sensor/pcnt_esp32/Kconfig"
 
+source "drivers/sensor/am2301b/Kconfig"
```
Comment (MaureenHelm) [inline drivers/sensor/Kconfig:261]: Please move this up to for alpha order

File: drivers/sensor/am2301b/am2301b.c [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+/**
+ * @brief Calculate the temperature value from the raw temperature value obtained from the sensor
+ *        Calculate the temperature from the formula in Chapter 6.2 of the AM2301B Product Manual
+ * @param val Temperature value after conversion
+ * @param temperature_raw Raw value of temperature obtained from the sensor(Sr)
+ */
+static inline void am2301b_convert_temperature_value(struct sensor_value *val,
+						     uint32_t temperature_raw)
+{
+	const static int32_t power20_of_2 = 0x100000;
```
Comment (MaureenHelm) [inline drivers/sensor/am2301b/am2301b.c:125]:
```suggestion
	const static int32_t power20_of_2 = BIT(20);
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Some minor nitpicks, but otherwise looks great. Thank you!

File: drivers/sensor/CMakeLists.txt (hunk 1 above)
Comment (yamaoku-009) [reply drivers/sensor/CMakeLists.txt]: fixed it

File: drivers/sensor/Kconfig (hunk 2 above)
Comment (yamaoku-009) [reply drivers/sensor/Kconfig]: fixed it.

File: drivers/sensor/am2301b/am2301b.c (hunk 3 above)
Comment (yamaoku-009) [reply drivers/sensor/am2301b/am2301b.c]: fixed it.

**Comment by yamaoku-009** [comment]
rebase

File: dts/bindings/sensor/aosong,am2301b.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022 Hisanori YAMAOKU
+# SPDX-License-Identifier: Apache-2.0
+description: Aosong AM2301B Humidity and Temperature Module
+compatible: "aosong,am2301b"
+
+include: i2c-device.yaml
```
Comment (MaureenHelm) [inline dts/bindings/sensor/aosong,am2301b.yaml:6]:
```suggestion
include: [sensor-device,yaml, i2c-device.yaml]
```

File: drivers/sensor/am2301b/am2301b.c [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+	.sample_fetch = am2301b_sample_fetch,
+	.channel_get = am2301b_channel_get,
+};
+
+#define AM2301B_INST(inst)						      \
+	static struct am2301b_data am2301b_data_##inst;			      \
+	static const struct am2301b_config am2301b_config_##inst = {	      \
+		.i2c = I2C_DT_SPEC_INST_GET(inst),			      \
+	};								      \
+	DEVICE_DT_INST_DEFINE(inst, am2301b_init, NULL, &am2301b_data_##inst, \
```
Comment (MaureenHelm) [inline drivers/sensor/am2301b/am2301b.c:225]:
```suggestion
	SENSOR_DEVICE_DT_INST_DEFINE(inst, am2301b_init, NULL, &am2301b_data_##inst, \
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
Need to rebase and resolve merge conflicts

**Comment by yamaoku-009** [comment]
Fixed merge conflicts

File: drivers/sensor/am2301b/Kconfig [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2022 Hisanori YAMAOKU
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_AOSONG_AM2301B := aosong,am2301b
+
+menuconfig AM2301B
+	bool "Aosong AM2301B Humidity and Temperature Module"
+	depends on I2C
```
Comment (MaureenHelm) [inline drivers/sensor/am2301b/Kconfig:8]:
```suggestion
	select I2C
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/am2301b/Kconfig (hunk 6 above)
Comment (yamaoku-009) [reply drivers/sensor/am2301b/Kconfig]: fixed it.

**Review by MaureenHelm — DISMISSED** [review]
Needs another rebase

**Review by avisconti — DISMISSED** [review]

**Comment by yamaoku-009** [comment]
Fixed merge conflicts

**Review by MaureenHelm — DISMISSED** [review]

**Comment by yamaoku-009** [comment]
Fixed merge conflicts

**Comment by MaureenHelm** [comment]
Needs rebase

**Comment by yamaoku-009** [comment]
Fixed merge conflicts

File: drivers/sensor/am2301b/Kconfig [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2022 Hisanori YAMAOKU
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_AOSONG_AM2301B := aosong,am2301b
```
Comment (MaureenHelm) [inline drivers/sensor/am2301b/Kconfig:4]: This isn't needed, please remove

File: drivers/sensor/am2301b/am2301b.c [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+	}
+}
+
+static int am2301b_init(const struct device *dev)
+{
+	const struct am2301b_config *cfg = dev->config;
+	uint8_t status;
+	int ret;
+
+	if (!device_is_ready(cfg->i2c.bus)) {
```
Comment (MaureenHelm) [inline drivers/sensor/am2301b/am2301b.c:180]: Use `i2c_is_ready_dt`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Needs another rebase

File: drivers/sensor/am2301b/Kconfig (hunk 7 above)
Comment (yamaoku-009) [reply drivers/sensor/am2301b/Kconfig]: fixed it.

File: drivers/sensor/am2301b/am2301b.c (hunk 8 above)
Comment (yamaoku-009) [reply drivers/sensor/am2301b/am2301b.c]: fixed it.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by carlescufi** [comment]
@yamaoku-009 please rebase to resolve the conflict

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
@teburd @yperess will you please take a look?

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 9]
```diff
@@ -794,3 +794,9 @@ test_i2c_adxl367: adxl367@77 {
 	reg = <0x77>;
 	odr = <4>;
 };
+
+test_i2c_am2301b: am2301b@78 {
+	compatible = "aosong,am2301b";
+	reg = <0x78>;
+	status = "okay";
+};
```
Comment (kartben) [inline tests/drivers/build_all/sensor/i2c.dtsi:802]: will need a rebase, 0x79 at the time of writing this

```suggestion
test_i2c_am2301b: am2301b@79 {
	compatible = "aosong,am2301b";
	reg = <0x79>;
	status = "okay";
};
```

File: drivers/sensor/am2301b/Kconfig [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+	  Enable driver for Aosong AM2301B Humidity and Temperature Module.
+
+if AM2301B
+
+config AM2301B_CRC_CHECK
+	bool "Use CRC Check"
+	imply CRC
+	default y
+	help
+	  Check the crc of measument data.
```
Comment (kartben) [inline drivers/sensor/am2301b/Kconfig:19]:
```suggestion
	  Check the crc of measurement data.
```

File: drivers/sensor/am2301b/am2301b.c [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+struct am2301b_config {
+	struct i2c_dt_spec i2c;
+};
+
+/* Initialization command from Chapter 5.4, Section 1 of AM2301B Product Manuals */
+const static uint8_t am2301b_cmd_initialize[] = { 0xBE, 0x08, 0x00 };
+/* Measurement trigger command from Chapter 5.4, Section 2 of AM2301B Product Manuals */
+const static uint8_t am2301b_cmd_measurement[] = { 0xAC, 0x33, 0x00 };
+
+static int am2301b_fetch_measument_data(const struct am2301b_config *cfg,
```
Comment (kartben) [inline drivers/sensor/am2301b/am2301b.c:52]:
```suggestion
static int am2301b_fetch_measurement_data(const struct am2301b_config *cfg,
```

File: drivers/sensor/am2301b/am2301b.c [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+	const struct am2301b_config *cfg = dev->config;
+	struct am2301b_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_ALL:
+		__fallthrough;
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		__fallthrough;
+	case SENSOR_CHAN_HUMIDITY:
+		return am2301b_fetch_measument_data(cfg, data);
```
Comment (kartben) [inline drivers/sensor/am2301b/am2301b.c:110]:
```suggestion
		return am2301b_fetch_measurement_data(cfg, data);
```

**Review by kartben — CHANGES_REQUESTED** [review]
spotted a typo + you need to rebase due to new i2c sensor recently added to the build_all devicetree. Thanks!

**Review by teburd — DISMISSED** [review]

**Comment by kartben** [comment]
@yamaoku-009 any chance you could address the typo highlighted in my earlier comment? Thank you in advance!

**Review by kartben — DISMISSED** [review]

**Comment by yamaoku-009** [comment]
Fixed merge conflicts.

**Comment by MaureenHelm** [comment]
Superseded by #67279

**Pull request closed without merging**